server/etcdserver/api/etcdhttp/health.go

```diff
@@ -91,6 +98,32 @@ func NewHealthHandler(lg *zap.Logger, hfunc func(excludedAlarms AlarmSet, Serial
 	}
 }
 
+// newHealthHandler generates a http HandlerFunc for a health check function hfunc.
+func newHealthHandler(path string, lg *zap.Logger, hfunc func(*http.Request) (Health, error)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			w.Header().Set("Allow", http.MethodGet)
+			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+			lg.Warn("/health error", zap.Int("status-code", http.StatusMethodNotAllowed))
+			return
+		}
+		h, err := hfunc(r)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
Returning BadRequest here seems incorrect. See line 272. `runHealthChecks` might return `fmt.Errorf("Health check: %s not registered", checkName)` which is a programmer error not a client error like `StatusBadRequest` https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400.

`health check not registered` error can only happen if `CheckRegistry` has a bug that would cause `checks` and `checkNames` to be inconsistent. This is a invariant of `CheckRegistry` as so it should never normally happen, but when it does we are screwed. 

I would definitely not attribute this to client error, and I would consider `CheckRegistry` panic instead.

Reply:
changed to panic.